src/payment/strategies/stripe-upe/stripe-upe-payment-strategy.ts

```diff
@@ -0,0 +1,189 @@
+
+        if (useStoreCredit) {
+            await this._store.dispatch(this._storeCreditActionCreator.applyStoreCredit(useStoreCredit));
+        }
+
+        try {
+            const paymentMethod = this._store.getState().paymentMethods.getPaymentMethodOrThrow(methodId);
+            const { clientToken } = paymentMethod;
+            if (!clientToken) {
+                throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+            }
+            const elements = this._getStripeElements();
+
+            const { paymentIntent, error } = await this._getStripeJs().confirmPayment({
+                elements,
+                redirect: 'if_required',
+            });
+
+            if (error) {
+                if (error.type === 'card_error' || error.type === 'invalid_request_error' || error.type === 'validation_error') {
+                    throw new Error(error.message);
+                }
+                throw new RequestError();
+            }
+
+            await this._store.dispatch(this._orderActionCreator.submitOrder(order, options));
+
+            const paymentPayload = {
+                methodId,
+                paymentData: {
+                    formattedPayload: {
+                        credit_card_token: { token: paymentIntent ? paymentIntent.id : '' },
+                        vault_payment_instrument: shouldSaveInstrument,
+                        confirm: false,
+                    },
+                    shouldSetAsDefaultInstrument,
+                },
+            };
+
+            return await this._store.dispatch(this._paymentActionCreator.submitPayment(paymentPayload));
+        } catch (error) {
+            throw error;
+        }
+    }
+
+    finalize(): Promise<InternalCheckoutSelectors> {
+        return Promise.reject(new OrderFinalizationNotRequiredError());
+    }
+
+    deinitialize(): Promise<InternalCheckoutSelectors> {
+        this._unmountElement();
+
+        return Promise.resolve(this._store.getState());
+    }
+
+    private _getInitializeOptions(): StripeUPEPaymentInitializeOptions {
+        if (!this._initializeOptions) {
+            throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
+        }
+
+        return this._initializeOptions;
+    }
+
+    private _getStripeElements(): StripeElements {
+        if (!this._stripeElements) {
+            throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
+        }
+
+        return this._stripeElements;
+    }
+
+    private _getStripeJs(): StripeUPEClient {
+        if (!this._stripeUPEClient) {
+            throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
+        }
+
+        return this._stripeUPEClient;
+    }
+
+    private async _loadStripeJs(stripePublishableKey: string, stripeConnectedAccount: string): Promise<StripeUPEClient> {
+        if (this._stripeUPEClient) { return Promise.resolve(this._stripeUPEClient); }
```

Comment:
This is a bit of a strange pattern, we check in the method to see if `_stripeUPEClient` is defined, but we set it in the parent. a better approach would be to do everything within the `_getStripeJs` method.

A much better approach would be to do all the caching within the `_getStripeJs` method:


```ts
_getStripeJs() {
  if (!this._stripeUPEClient) {
    this._stripeUPEClient = this._stripeScriptLoader.load(
      stripePublishableKey,
      stripeConnectedAccount,
      this._locale
    );
  }

  return this._stripeUPEClient;
}
```

With that in place, we can just use `_getStripeJs` in `_mountCardFields`:

```ts
    private _mountCardFields(clientToken: string = ''): Promise<StripeElement> {
        const { containerId } = this._getInitializeOptions();

        const elementOptions: StripeElementsOptions = {
            clientSecret: clientToken,
        };

        const stripeJs = await this._getStripeJs();

        return new Promise((resolve, reject) => {
            if (!this._stripeElements) {
                this._stripeElements = stripeJs.elements(elementOptions);
            }

            const stripeElement = this._stripeElements.getElement('payment') || this._stripeElements.create('payment',
            // ...
```

what do you think?